### ISSUE TITLE
Training model with language Vietnamese

### ISSUE THREAD

**Issue by minhduc01168** [YutingLi0606/HTR-VT#7]
I have edited the code a bit to implement training model for Vietnamese language. However, I feel that the training process is not correct. Because the val loss is too large and CER and WER are both equal to 1 and not decreasing? Can you advise me on the problem I'm having? I'm very grateful for that.
The training results on my 3000 sample dataset are as follows:
2025-02-12 04:27:32,899 INFO Loading train loader...
2025-02-12 04:27:32,905 INFO Loading val loader...
2025-02-12 04:27:52,938 INFO Epoch 1     Training Loss: 259.3788
2025-02-12 04:28:01,286 INFO New best CER: 1.0000 (previous inf)
2025-02-12 04:28:01,561 INFO New best WER: 1.0000 (previous inf)
2025-02-12 04:28:01,836 INFO Epoch 1     Val Loss: 630.9618      CER: 1.0000     WER: 1.0000
2025-02-12 04:28:22,355 INFO Epoch 2     Training Loss: 25.5186
2025-02-12 04:28:29,691 INFO Epoch 2     Val Loss: 629.7071      CER: 1.0000     WER: 1.0000
2025-02-12 04:28:51,749 INFO Epoch 3     Training Loss: 23.4880
2025-02-12 04:28:58,956 INFO Epoch 3     Val Loss: 627.9271      CER: 1.0000     WER: 1.0000
2025-02-12 04:29:17,199 INFO Epoch 4     Training Loss: 21.0262
2025-02-12 04:29:24,144 INFO Epoch 4     Val Loss: 625.9325      CER: 1.0000     WER: 1.0000
2025-02-12 04:29:45,714 INFO Epoch 5     Training Loss: 20.1412
2025-02-12 04:29:53,510 INFO Epoch 5     Val Loss: 623.8199      CER: 1.0000     WER: 1.0000
2025-02-12 04:30:11,689 INFO Epoch 6     Training Loss: 19.5137
2025-02-12 04:30:19,286 INFO Epoch 6     Val Loss: 621.6227      CER: 1.0000     WER: 1.0000
2025-02-12 04:30:37,859 INFO Epoch 7     Training Loss: 19.0466
2025-02-12 04:30:45,053 INFO Epoch 7     Val Loss: 619.1792      CER: 1.0000     WER: 1.0000
2025-02-12 04:31:06,211 INFO Epoch 8     Training Loss: 18.7107
2025-02-12 04:31:13,994 INFO Epoch 8     Val Loss: 616.6867      CER: 1.0000     WER: 1.0000
2025-02-12 04:31:36,924 INFO Epoch 9     Training Loss: 18.4415
2025-02-12 04:31:44,057 INFO Epoch 9     Val Loss: 614.2237      CER: 1.0000     WER: 1.0000
2025-02-12 04:32:04,858 INFO Epoch 10    Training Loss: 18.1325
2025-02-12 04:32:13,625 INFO Epoch 10    Val Loss: 611.4538      CER: 1.0000     WER: 1.0000
2025-02-12 04:32:34,764 INFO Epoch 11    Training Loss: 17.9061
2025-02-12 04:32:42,000 INFO Epoch 11    Val Loss: 609.0321      CER: 1.0000     WER: 1.0000
2025-02-12 04:33:03,048 INFO Epoch 12    Training Loss: 17.6806
2025-02-12 04:33:11,003 INFO Epoch 12    Val Loss: 606.1822      CER: 1.0000     WER: 1.0000
2025-02-12 04:33:32,113 INFO Epoch 13    Training Loss: 17.4921
2025-02-12 04:33:40,182 INFO Epoch 13    Val Loss: 603.3437      CER: 1.0000     WER: 1.0000
2025-02-12 04:34:02,338 INFO Epoch 14    Training Loss: 17.3500
2025-02-12 04:34:10,579 INFO Epoch 14    Val Loss: 600.5021      CER: 1.0000     WER: 1.0000
2025-02-12 04:34:30,764 INFO Epoch 15    Training Loss: 17.1961
2025-02-12 04:34:38,433 INFO Epoch 15    Val Loss: 597.3107      CER: 1.0000     WER: 1.0000
2025-02-12 04:34:57,064 INFO Epoch 16    Training Loss: 16.9961
2025-02-12 04:35:04,185 INFO Epoch 16    Val Loss: 594.6144      CER: 1.0000     WER: 1.0000
2025-02-12 04:35:26,303 INFO Epoch 17    Training Loss: 16.9804
2025-02-12 04:35:33,435 INFO Epoch 17    Val Loss: 591.2913      CER: 1.0000     WER: 1.0000
2025-02-12 04:35:54,403 INFO Epoch 18    Training Loss: 16.8826
2025-02-12 04:36:02,307 INFO Epoch 18    Val Loss: 587.7942      CER: 1.0000     WER: 1.0000
2025-02-12 04:36:20,644 INFO Epoch 19    Training Loss: 16.7043
2025-02-12 04:36:28,891 INFO Epoch 19    Val Loss: 584.3790      CER: 1.0000     WER: 1.0000
2025-02-12 04:36:47,349 INFO Epoch 20    Training Loss: 16.6062
2025-02-12 04:36:55,572 INFO Epoch 20    Val Loss: 580.8413      CER: 1.0000     WER: 1.0000
2025-02-12 04:37:13,774 INFO Epoch 21    Training Loss: 16.5109
2025-02-12 04:37:21,941 INFO Epoch 21    Val Loss: 577.1611      CER: 1.0000     WER: 1.0000
2025-02-12 04:37:44,070 INFO Epoch 22    Training Loss: 16.4650
2025-02-12 04:37:51,135 INFO Epoch 22    Val Loss: 573.8587      CER: 1.0008     WER: 1.0000
2025-02-12 04:38:13,126 INFO Epoch 23    Training Loss: 16.3583
2025-02-12 04:38:20,393 INFO Epoch 23    Val Loss: 570.0486      CER: 1.0024     WER: 1.0000
2025-02-12 04:38:40,622 INFO Epoch 24    Training Loss: 16.2070
2025-02-12 04:38:48,659 INFO Epoch 24    Val Loss: 565.7410      CER: 1.0043     WER: 1.0000
2025-02-12 04:39:07,807 INFO Epoch 25    Training Loss: 16.0739
2025-02-12 04:39:15,068 INFO Epoch 25    Val Loss: 562.0154      CER: 1.0091     WER: 1.0000
2025-02-12 04:39:38,752 INFO Epoch 26    Training Loss: 16.1597
2025-02-12 04:39:46,529 INFO Epoch 26    Val Loss: 557.0933      CER: 1.0178     WER: 1.0000
2025-02-12 04:40:09,385 INFO Epoch 27    Training Loss: 15.9085
2025-02-12 04:40:17,462 INFO Epoch 27    Val Loss: 552.5342      CER: 1.0276     WER: 1.0000
2025-02-12 04:40:37,863 INFO Epoch 28    Training Loss: 15.6608
2025-02-12 04:40:46,940 INFO Epoch 28    Val Loss: 547.9150      CER: 1.0399     WER: 1.0000
2025-02-12 04:41:08,978 INFO Epoch 29    Training Loss: 15.4818
2025-02-12 04:41:16,253 INFO Epoch 29    Val Loss: 543.8703      CER: 1.0944     WER: 1.0000
2025-02-12 04:41:36,099 INFO Epoch 30    Training Loss: 15.2258
2025-02-12 04:41:44,528 INFO Epoch 30    Val Loss: 538.4594      CER: 1.1180     WER: 1.0000
2025-02-12 04:42:04,409 INFO Epoch 31    Training Loss: 15.1075
2025-02-12 04:42:12,697 INFO Epoch 31    Val Loss: 533.7211      CER: 1.1717     WER: 1.0000
2025-02-12 04:42:33,098 INFO Epoch 32    Training Loss: 14.9345
2025-02-12 04:42:40,811 INFO Epoch 32    Val Loss: 528.6767      CER: 1.2606     WER: 1.0000
2025-02-12 04:43:01,139 INFO Epoch 33    Training Loss: 14.7049
2025-02-12 04:43:08,349 INFO Epoch 33    Val Loss: 523.8678      CER: 1.3143     WER: 1.0000

**Comment by daominhwysi** [comment]
Have you adjusted the number of characters for the CTC loss?

**Comment by johnlockejrr** [comment]
> Have you adjusted the number of characters for the CTC loss?

How you do that?

**Comment by YutingLi0606** [comment]
Hi，thanks for your interest~
1. I'd suggest paying attention to the character number of your datasets and ensuring it aligns with the number of classes in your CTC loss configuration.

2. HTR-VT utilizes a slightly modified ResNet-18 instead of the standard Patch Embedding in ViT. The downsampling size depends on this ResNet configuration. If you wish to use a different image size, you will need to adjust the ResNet accordingly.

In options.py, make sure to align the patch size with the new downsampling size.

For example, we use an input size of 512x64 with a downsampling size of 4x64 in ResNet (corresponding to the patch size in options.py). This results in a 128x1 feature input to the transformer encoder. Based on our experiments, maintaining this feature shape as nx1 is optimal for this task.

If you plan to use a size like 150x40, here are two suggestions:

Consider resizing the input to 128x32 or using a higher resolution like 256x64.
Modify the ResNet (adjust the stride, add layers, or remove layers) to ensure the downsampling size remains appropriate.
For instance, with an input size of 128x32, after adjustments, you can obtain a feature size of 32x1. In this case, the patch size would be 4x32. Modify the ResNet configuration as follows:
self.layer2 = self._make_layer(BasicBlock, nb_feat // 2, 2, stride=2)
Change to:
self.layer2 = self._make_layer(BasicBlock, nb_feat // 2, 2, stride=(2, 1))

Note: The hyperparameters provided above are examples. You should fine-tune them based on your experimental results.

3. From your training logs, it seems there might be an issue in your validation pipeline. You should check some details(e.g., try printing key variables). 

I hope this helps!

Best,
Yuting
